### PR TITLE
Use SmallVec for SmallCStr

### PR DESCRIPTION
**Pull request by llogiq**
This reuses the awesome optimizations from Servo's `SmallVec` to speed up `SmallCStr`.

### CODE REVIEW

File: src/librustc_data_structures/small_c_str.rs [hunk 1]
```diff
@@ -11,69 +11,61 @@
 use std::ffi;
 use std::ops::Deref;
 
-const SIZE: usize = 38;
+use smallvec::SmallVec;
+
+const SIZE: usize = 36;
```
Comment (estebank) [inline src/librustc_data_structures/small_c_str.rs]: Any reason for the change?

File: src/librustc_data_structures/small_c_str.rs [hunk 2]
```diff
@@ -11,69 +11,61 @@
...
-            SmallCStr::OnHeap {
-                data: ffi::CString::new(s).unwrap()
-            }
+            let mut data = Vec::with_capacity(len1);
+            data.extend_from_slice(s.as_bytes());
+            data.push(0);
+            SmallVec::from_vec(data)
+        };
+        if let Err(e) = ffi::CStr::from_bytes_with_nul(&data) {
+            panic!("The string \"{}\" cannot be converted into a CStr: {}", s, e);
```
Comment (estebank) [inline src/librustc_data_structures/small_c_str.rs]: I know you're not introducing this, but should we be returning `Result` here instead? Probably ok given the usage, but still...

File: src/librustc_data_structures/small_c_str.rs [hunk 3]
```diff
@@ -11,69 +11,61 @@
...
-pub enum SmallCStr {
-    OnStack {
-        data: [u8; SIZE],
-        len_with_nul: u8,
-    },
-    OnHeap {
-        data: ffi::CString,
-    }
+pub struct SmallCStr {
+    data: SmallVec<[u8; SIZE]>,
```
Comment (estebank) [inline src/librustc_data_structures/small_c_str.rs]: Ah! Now I see what you're doing. Why is this an improvement over the `OnHeap` version?

**Review by estebank — APPROVED** [review]

**Comment by estebank** [comment]
r=me, but I wan't to know the response for the comments

File: src/librustc_data_structures/small_c_str.rs [hunk 4]
```diff
@@ -11,69 +11,61 @@
...
+            SmallVec::from_vec(data)
+        };
+        if let Err(e) = ffi::CStr::from_bytes_with_nul(&data) {
+            panic!("The string \"{}\" cannot be converted into a CStr: {}", s, e);
         }
+        SmallCStr { data }
     }
 
+    #[inline]
+    pub fn new_with_zero(s: &str) -> SmallCStr {
```
Comment (nicokoch) [inline src/librustc_data_structures/small_c_str.rs]: This should probably be called `new_with_nul` , or `from_str_with_nul`.

File: src/librustc_data_structures/small_c_str.rs (hunk 2 above)
Comment (llogiq) [reply src/librustc_data_structures/small_c_str.rs]: This was already discussed during the introduction of the SmallCStr type, the consensus was that ergonomics are more important as long as it's possible to get a stack trace.

File: src/librustc_data_structures/small_c_str.rs (hunk 1 above)
Comment (llogiq) [reply src/librustc_data_structures/small_c_str.rs]: Yes, `Array` isn't implemented for `[T; 38]`, and due to the union optimization, the on-stack version can take more content per byte.

File: src/librustc_data_structures/small_c_str.rs (hunk 3 above)
Comment (llogiq) [reply src/librustc_data_structures/small_c_str.rs]: Absent the union optimization, this should generate the same code. However *with* the union optimization, the type is reduced to `(usize, [T; _] / *mut T)`, which saves the discriminant + padding.

Apart from the memory savings, the code has seen a good deal of perf work already.

File: src/librustc_data_structures/small_c_str.rs (hunk 4 above)
Comment (llogiq) [reply src/librustc_data_structures/small_c_str.rs]: Thanks, changed.

**Pull request merged**